### PR TITLE
Add command to generate `.sdkmanrc` file

### PR DESCRIPTION
**Pull request by helpermethod**
To raise a new Pull Request, the following prerequisites need to be met. Please tick before proceeding:

- [x] a conversation was held in the appropriate [SDKMAN Slack](https://slack.sdkman.io) channel.
- [x] a Github Issue was opened for this feature / bug.
- [x] test coverage was added (Cucumber or Spock as appropriate).

This addresses Issue #730

### CODE REVIEW

**Comment by helpermethod** [comment]
Hi @marc0der,

this is the current WIP of the `sdk env init` command.

Some open questions:

* Which Java version should I add as an example? Is there some easy way to get the currently used Java version, and fall back on some default?
* If `sdk env` is called with some "subcommand" other than `init` what would be the expected outcome? Should an error be displayed? Should just `sdk env` be called?

File: src/main/bash/sdkman-env.sh [hunk 1]
```diff
@@ -30,14 +30,26 @@ function __sdk_env() {
 	local line_number=0
 
 	while IFS= read -r line || [[ -n $line ]]; do
-		if [[ ! $line =~ ^[[:lower:]]+=.+$ ]]; then
-			__sdkman_echo_red "${sdkmanrc}:${line_number}: Invalid candidate format! Expected 'candidate=version' but found '$line'"
+		__sdkman_is_blank_or_comment "$line" && continue
+
+		if ! __sdkman_matches_candidate_format "$line"; then
+			__sdkman_echo_red "${sdkrc}:${line_number}: Invalid candidate format! Expected 'candidate version' but found '$line'"
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh]: Can we split this into two lines please, separated by a newline .
The warning in red, the further information in yellow.

File: src/main/bash/sdkman-env.sh [hunk 2]
```diff
@@ -30,14 +30,26 @@ function __sdk_env() {
...
+		local candidate version
+		IFS=' \t' read -r candidate version <<< "$line"
+		__sdk_use "$candidate" "$version"
 
 		((line_number++))
-	done < "$sdkmanrc"
+	done < "$sdkrc"
+}
+
+function __sdkman_is_blank_or_comment() {
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh]: how about `__sdkman_line_is_blank_or_comment()`?

File: src/main/bash/sdkman-env.sh [hunk 3]
```diff
@@ -30,14 +30,26 @@ function __sdk_env() {
...
+	[[ $1 =~ ^[[:blank:]]*|\#.*$ ]]
 }
+
+function __sdkman_matches_candidate_format() {
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:58]: `__sdkman_line_matches_candidate_format()`?

File: src/main/bash/sdkman-env.sh [hunk 4]
```diff
@@ -30,26 +30,26 @@ function __sdk_env() {
...
+		IFS=' \t' read -r candidate version rest <<< "$line"
 		__sdk_use "$candidate" "$version"
 
 		((line_number++))
 	done < "$sdkrc"
 }
 
-function __sdkman_is_blank_or_comment() {
-	[[ $1 =~ ^[[:blank:]]*|\#.*$ ]]
+function __sdkman_is_blank_line_or_comment_line() {
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh]: Lol, don't worry. I see it got addressed in a later commit.

File: src/main/bash/sdkman-env.sh [hunk 5]
```diff
@@ -17,10 +17,10 @@
 #
 
 function __sdk_env() {
-	readonly sdkrc='.sdkrc'
+	readonly sdkmanrc='.sdkmanrc'
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:20]: I like this better ;-)

File: src/main/bash/sdkman-env.sh [hunk 6]
```diff
@@ -19,30 +19,35 @@
...
+		cat <<- EOF > "$sdkmanrc"
+		# This file was created by SDKMAN.
+		# 
+		# java=8.0.252-zulu 
+		EOF
+	fi
+
 	if [[ ! -f "$sdkmanrc" ]]; then
-		__sdkman_echo_red "No $sdkmanrc file found."
+		__sdkman_echo_red "SDKMAN can't find an .sdkmanrc file in your current directory."
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:31]: How about _Could not find .sdkmanrc file in current directory._?

File: src/main/bash/sdkman-env.sh [hunk 7]
```diff
@@ -19,30 +19,35 @@
...
 		echo ""
-		__sdkman_echo_yellow "Please create one before using this command."
+		__sdkman_echo_yellow "We recommend creating one by entering 'sdk env init'."
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:33]: _Run 'sdk env init' to create it._

**Comment by marc0der** [comment]
Apologies, I was reviewing this commit by commit so not all the comments make sense by the end.

**Comment by helpermethod** [comment]
@marc0der Thanks for the valuable suggestions, I suck at error messages / warnings ;).

I've also simplified things a bit, the lines are now normalised (stripped by whitespace and comments) before performing the format check, which will help the user in case he entered something invalid.

I've also fallen back to using `=` for separating candidate and version, as some devs may already use the beta version (and it actually doesn't make a difference what delimiter is used).

**Comment by marc0der** [comment]
Sounds sensible yes. Haven't looked into it, but the build on your branch seems to be failing. Once we have a green build I'd be happy to merge.

**Comment by helpermethod** [comment]
Hi @marc0der,

I've fixed the outstanding issues, added some more tests, fixed some broken tests and changed how the `.sdkmanrc` file is created (it inserts the currently used java version or falls back to some fixed java version).

There are still some untested code branches and minor things which can be improved but IMHO nothing which prevents a release.

**Comment by helpermethod** [comment]
It seems I can use `version=​$(​__sdkman_secure_curl ​"​${SDKMAN_CANDIDATES_API}​/candidates/default/​${candidate}​"​)​` to determine the fallback Java version?

File: src/test/groovy/sdkman/specs/EnvCommandSpec.groovy [hunk 8]
```diff
@@ -24,7 +31,7 @@ class EnvCommandSpec extends SdkmanEnvSpecification {
 			}
 		}
 
-		new File(bash.workDir, '.sdkmanrc').text = sdkmanrc
+		new File(bash.workDir, '.sdkmanrc').text = sdkrc
```
Comment (marc0der) [inline src/test/groovy/sdkman/specs/EnvCommandSpec.groovy:34]: Should this be `sdkmanrc`?

File: src/test/groovy/sdkman/specs/EnvCommandSpec.groovy [hunk 9]
```diff
@@ -36,20 +43,50 @@ class EnvCommandSpec extends SdkmanEnvSpecification {
 		}
 
 		where:
-		sdkmanrc << ["grails=2.1.0\ngroovy=2.4.1", "grails=2.1.0\ngroovy=2.4.1\n"]
+		sdkrc << [
```
Comment (marc0der) [inline src/test/groovy/sdkman/specs/EnvCommandSpec.groovy:46]: Same as above.

File: src/test/groovy/sdkman/specs/EnvCommandSpec.groovy [hunk 10]
```diff
@@ -36,20 +43,50 @@ class EnvCommandSpec extends SdkmanEnvSpecification {
...
+		new File(bash.workDir, ".sdkmanrc").text = sdkrc
+
+		when:
+		bash.execute("sdk env")
+
+		then:
+		bash.output.contains("Using groovy version 2.4.1 in this shell.")
+
+		where:
+		sdkrc << [
```
Comment (marc0der) [inline src/test/groovy/sdkman/specs/EnvCommandSpec.groovy:86]: It is just naming, but let's call it `sdkmanrc` throughout.

**Comment by marc0der** [comment]
Some very minor naming bits  in the tests, but almost ready to merge! Also, I hope you don't mind me squashing your commits.

File: src/test/groovy/sdkman/specs/EnvCommandSpec.groovy (hunk 10 above)
Comment (helpermethod) [reply src/test/groovy/sdkman/specs/EnvCommandSpec.groovy]: Consistency is important, I agree.

**Comment by helpermethod** [comment]
Hi @marc0der,

I've added one more test and performed the suggested fixes.

File: src/main/bash/sdkman-env.sh [hunk 11]
```diff
@@ -17,27 +17,67 @@
 #
 
 function __sdk_env() {
-	readonly sdkmanrc='.sdkmanrc'
+	SDKMANRC=".sdkmanrc"
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:20]: Was there a problem with `readonly`? I would even be happy with it being `local`, but don't really like it defined as global though.

File: src/main/bash/sdkman-env.sh [hunk 12]
```diff
@@ -17,27 +17,67 @@
...
 
-	if [[ ! -f "$sdkmanrc" ]]; then
-		__sdkman_echo_red "No $sdkmanrc file found."
+	if [[ "$1" == "init" ]]; then
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:22]: What exactly is `"$1"` here? I would suggest assigning whatever it is to a local variable and then using that here instead.

File: src/main/bash/sdkman-env.sh [hunk 13]
```diff
@@ -17,27 +17,67 @@
...
+
+	echo "java=$version" > "$SDKMANRC"
+
+	__sdkman_echo_green "$SDKMANRC created."
+}
 
-		((line_number++))
-	done < "$sdkmanrc"
+function __sdkman_is_blank_line() {
+	[[ -z "$1" ]]
```
Comment (marc0der) [inline src/main/bash/sdkman-env.sh:72]: What will this return? An exit code?

File: src/main/bash/sdkman-env.sh (hunk 11 above)
Comment (rgoldberg) [reply src/main/bash/sdkman-env.sh:20]: I don't know why @helpermethod mad it a global.  `readonly` variables fail in bash 3 when a method is called more than once, as `readonly` in bash 3 creates a readonly global.  The way around this is to use `local -r` which works in bash 3, as well as in bash 5 & in zsh.  If that's why he changed it, using `local -r` should fix the problem.

Comment (helpermethod) [reply src/main/bash/sdkman-env.sh:20]: I will make it local and pass it explicitly to the _generate function.

File: src/main/bash/sdkman-env.sh (hunk 13 above)
Comment (helpermethod) [reply src/main/bash/sdkman-env.sh:72]: From the bash manpage

> When executed, the exit status of a function is the exit status of the last command executed in the body.

File: src/main/bash/sdkman-env.sh (hunk 11 above)
Comment (helpermethod) [reply src/main/bash/sdkman-env.sh]: I've mad the variable local and readonly (local - r).

File: src/main/bash/sdkman-env.sh (hunk 12 above)
Comment (helpermethod) [reply src/main/bash/sdkman-env.sh]: Assigned $1 to a variable.

**Pull request merged**